File: backend/app/schemas/rbac.py
from pydantic import BaseModel, Field

# ── Permission Tree ────────────────────────────────────────────────────────────


class PermissionGroupCreate(BaseModel):
    name: str = Field(max_length=100)
    code: str = Field(max_length=50)
    description: str | None = None
    sort_order: int = 0


class PermissionPageCreate(BaseModel):
    group_id: int
    name: str = Field(max_length=100)
    code: str = Field(max_length=50)
    page_url: str | None = None
    description: str | None = None
    sort_order: int = 0


class PermissionPublic(BaseModel):
    id: int
    name: str
    code: str
    full_code: str
    is_builtin: bool

    model_config = {"from_attributes": True}


class PermissionPagePublic(BaseModel):
    id: int
    name: str
    code: str
    page_url: str | None
    sort_order: int
    permissions: list[PermissionPublic] = []

    model_config = {"from_attributes": True}


class PermissionGroupPublic(BaseModel):
    id: int
    name: str
    code: str
    sort_order: int
    pages: list[PermissionPagePublic] = []

    model_config = {"from_attributes": True}


class PermissionTreeResponse(BaseModel):
    groups: list[PermissionGroupPublic]


# ── Role ───────────────────────────────────────────────────────────────────────


class RoleCreate(BaseModel):
    name: str = Field(max_length=100)
    code: str = Field(max_length=50)
    description: str | None = None


class RoleUpdate(BaseModel):
    name: str | None = None
    description: str | None = None


class RolePublic(BaseModel):
    id: int
    name: str
    code: str
    description: str | None
    is_builtin: bool

    model_config = {"from_attributes": True}


class RolesPublic(BaseModel):
    data: list[RolePublic]
    count: int


class RoleOptionPublic(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


class AssignRolePermissions(BaseModel):
    """Replaces all current permissions on a role."""
    permission_ids: list[int]


class AssignUserRoles(BaseModel):
    """Replaces all current roles for a user."""
    role_ids: list[int]


class AssignRoleUsers(BaseModel):
    """Replaces all current users assigned to a role."""
    user_ids: list[str]


# ── Navigation (user-specific view of permission tree) ─────────────────────────


class NavigationPermission(BaseModel):
    id: int
    name: str
    code: str
    full_code: str


class NavigationPage(BaseModel):
    id: int
    name: str
    code: str
    page_url: str | None
    sort_order: int
    permissions: list[NavigationPermission]


class NavigationGroup(BaseModel):
    id: int
    name: str
    code: str
    sort_order: int
    pages: list[NavigationPage]


class NavigationResponse(BaseModel):
    groups: list[NavigationGroup]

File: frontend/src/api/rbac.ts
import api from '@/api'
import type { UserSearchResults } from '@/api/users'

export interface RolePublic {
  id: number
  name: string
  code: string
  description: string | null
  is_builtin: boolean
}

export interface RolesPublic {
  data: RolePublic[]
  count: number
}

export interface PermissionPublic {
  id: number
  name: string
  code: string
  full_code: string
  is_builtin: boolean
}

export interface PermissionPagePublic {
  id: number
  name: string
  code: string
  page_url: string | null
  sort_order: number
  permissions: PermissionPublic[]
}

export interface PermissionGroupPublic {
  id: number
  name: string
  code: string
  sort_order: number
  pages: PermissionPagePublic[]
}

export interface PermissionTreeResponse {
  groups: PermissionGroupPublic[]
}

export const rolesApi = {
  list: () => api.get<RolesPublic>('/roles/'),
  get: (id: number) => api.get<RolePublic>(`/roles/${id}`),
  create: (data: { name: string; code: string; description?: string }) =>
    api.post<RolePublic>('/roles/', data),
  update: (
    id: number,
    data: { name?: string; description?: string; is_active?: boolean }
  ) => api.patch<RolePublic>(`/roles/${id}`, data),
  delete: (id: number) => api.delete(`/roles/${id}`),
  getPermissions: (id: number) => api.get<number[]>(`/roles/${id}/permissions`),
  assignPermissions: (id: number, permissionIds: number[]) =>
    api.put(`/roles/${id}/permissions`, { permission_ids: permissionIds }),
  getUsers: (id: number, params?: { keyword?: string }) =>
    api.get<UserSearchResults>(`/roles/${id}/users`, { params }),
  assignUsers: (id: number, userIds: string[]) =>
    api.put(`/roles/${id}/users`, { user_ids: userIds }),
}

export const permissionsApi = {
  tree: () => api.get<PermissionTreeResponse>('/permissions/tree'),
}
